File: go/push.go
package msgine

import (
	"context"
)

// PushService provides methods for sending push notifications and managing devices.
type PushService struct {
	t *transport
}

type sendPushRequest struct {
	Title          string            `json:"title"`
	Body           string            `json:"body"`
	UserIDs        []string          `json:"userIds,omitempty"`
	Topic          string            `json:"topic,omitempty"`
	Data           map[string]string `json:"data,omitempty"`
	ImageURL       string            `json:"imageUrl,omitempty"`
	ScheduledFor   string            `json:"scheduledFor,omitempty"`
	IdempotencyKey string            `json:"idempotencyKey,omitempty"`
}

// Send sends a push notification to a topic or specific users.
//
// Provide either Topic (broadcast) or UserIDs (targeted). If neither is set,
// the notification is sent to all devices registered under the authenticated account.
func (s *PushService) Send(ctx context.Context, opts SendPushOptions) (*PushMessage, error) {
	if opts.Title == "" {
		return nil, &ValidationError{Field: "title", Message: "title is required"}
	}
	if len(opts.Title) > 100 {
		return nil, &ValidationError{Field: "title", Message: "title must be 100 characters or fewer"}
	}
	if opts.Body == "" {
		return nil, &ValidationError{Field: "body", Message: "body is required"}
	}
	if len(opts.Body) > 500 {
		return nil, &ValidationError{Field: "body", Message: "body must be 500 characters or fewer"}
	}

	body := sendPushRequest{
		Title:          opts.Title,
		Body:           opts.Body,
		UserIDs:        opts.UserIDs,
		Topic:          opts.Topic,
		Data:           opts.Data,
		ImageURL:       opts.ImageURL,
		ScheduledFor:   opts.ScheduledFor,
		IdempotencyKey: opts.IdempotencyKey,
	}
	var result PushMessage
	if err := s.t.do(ctx, "POST", "/developers/push", nil, body, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// RegisterDevice registers an FCM device token so it can receive push notifications.
// Registering the same token again is safe (upsert).
func (s *PushService) RegisterDevice(ctx context.Context, opts RegisterDeviceOptions) (*DeviceRegistration, error) {
	if opts.Token == "" {
		return nil, &ValidationError{Field: "token", Message: "token is required"}
	}
	if opts.Platform == "" {
		return nil, &ValidationError{Field: "platform", Message: "platform is required"}
	}

	var result DeviceRegistration
	if err := s.t.do(ctx, "POST", "/developers/push/device", nil, opts, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// RemoveDevice deactivates a device token (e.g. on user logout).
func (s *PushService) RemoveDevice(ctx context.Context, token string) error {
	if token == "" {
		return &ValidationError{Field: "token", Message: "token is required"}
	}
	return s.t.do(ctx, "DELETE", "/developers/push/device", nil, map[string]string{"token": token}, nil)
}

// History returns a paginated list of sent push notifications.
func (s *PushService) History(ctx context.Context, page, limit int) (*PaginatedResult[PushMessage], error) {
	q := queryValues("page", page, "limit", limit)
	var result PaginatedResult[PushMessage]
	if err := s.t.do(ctx, "GET", "/developers/push/history", q, nil, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

File: src/modules/push.ts
import { HttpClient } from '../http-client';
import {
  DeviceRegistration,
  HttpMethod,
  MsGineValidationError,
  PaginationOptions,
  PushHistoryResult,
  PushMessage,
  RegisterDeviceOptions,
  RemoveDeviceOptions,
  SendPushOptions,
} from '../types';

export class PushModule {
  constructor(private readonly http: HttpClient) {}

  /**
   * Send a push notification to devices or an FCM topic.
   *
   * Devices must be registered via {@link registerDevice} first.
   *
   * @example
   * ```ts
   * // Send to specific users
   * await client.push.send({
   *   title: 'New message',
   *   body: 'You have a new message',
   *   userIds: ['user-uuid'],
   * });
   *
   * // Broadcast to a topic
   * await client.push.send({
   *   title: 'Platform update',
   *   body: 'Version 2.0 is live',
   *   topic: 'all-users',
   * });
   * ```
   */
  async send(options: SendPushOptions): Promise<PushMessage> {
    validate(options);

    return this.http.request<PushMessage>({
      method: HttpMethod.POST,
      path: '/developers/push',
      body: {
        title: options.title,
        body: options.body,
        userIds: options.userIds,
        topic: options.topic,
        data: options.data,
        imageUrl: options.imageUrl,
        scheduledFor: options.scheduledFor,
        idempotencyKey: options.idempotencyKey,
      },
    });
  }

  /**
   * Register an FCM device token so it can receive push notifications.
   *
   * Call this after your app obtains a Firebase registration token.
   * Registering the same token again is safe (upsert).
   *
   * @example
   * ```ts
   * // Web
   * import { getToken } from 'firebase/messaging';
   * const token = await getToken(messaging, { vapidKey: VAPID_KEY });
   * await client.push.registerDevice({ token, platform: 'web' });
   * ```
   */
  async registerDevice(options: RegisterDeviceOptions): Promise<DeviceRegistration> {
    if (!options.token) throw new MsGineValidationError('token is required', 'token');
    if (!options.platform) throw new MsGineValidationError('platform is required', 'platform');

    return this.http.request<DeviceRegistration>({
      method: HttpMethod.POST,
      path: '/developers/push/device',
      body: { token: options.token, platform: options.platform },
    });
  }

  /**
   * Deactivate a device token (e.g. on user logout).
   *
   * @example
   * ```ts
   * await client.push.removeDevice({ token: fcmToken });
   * ```
   */
  async removeDevice(options: RemoveDeviceOptions): Promise<void> {
    if (!options.token) throw new MsGineValidationError('token is required', 'token');

    await this.http.request<{ message: string }>({
      method: HttpMethod.DELETE,
      path: '/developers/push/device',
      body: { token: options.token },
    });
  }

  /**
   * Get paginated push notification history.
   */
  async history(options: PaginationOptions = {}): Promise<PushHistoryResult> {
    const raw = await this.http.request<{
      notifications: PushMessage[];
      total: number;
      page: number;
      pages: number;
    }>({
      method: HttpMethod.GET,
      path: '/developers/push/history',
      queryParams: { page: options.page ?? 1, limit: options.limit ?? 20 },
    });

    return { items: raw.notifications, total: raw.total, page: raw.page, pages: raw.pages };
  }
}

// ── Validation ────────────────────────────────────────────────────────────────

function validate(o: SendPushOptions): void {
  if (!o.title || o.title.trim().length === 0)
    throw new MsGineValidationError('title is required', 'title');
  if (o.title.length > 100)
    throw new MsGineValidationError('title must be 100 characters or fewer', 'title');
  if (!o.body || o.body.trim().length === 0)
    throw new MsGineValidationError('body is required', 'body');
  if (o.body.length > 500)
    throw new MsGineValidationError('body must be 500 characters or fewer', 'body');
  if (o.data) {
    for (const [k, v] of Object.entries(o.data)) {
      if (typeof v !== 'string')
        throw new MsGineValidationError(`data.${k} must be a string (FCM requirement)`, `data.${k}`);
    }
  }
}
